src/utils/loadFileHandler.js

```diff
@@ -0,0 +1,128 @@
+/* Copyright (c) 2015 - 2021, Nordic Semiconductor ASA
+ *
+ * All rights reserved.
+ *
+ * Use in source and binary forms, redistribution in binary form only, with
+ * or without modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions in binary form, except as embedded into a Nordic
+ *    Semiconductor ASA integrated circuit in a product or a software update for
+ *    such product, must reproduce the above copyright notice, this list of
+ *    conditions and the following disclaimer in the documentation and/or other
+ *    materials provided with the distribution.
+ *
+ * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
+ *    contributors may be used to endorse or promote products derived from this
+ *    software without specific prior written permission.
+ *
+ * 3. This software, with or without modification, must only be used with a Nordic
+ *    Semiconductor ASA integrated circuit.
+ *
+ * 4. Any software provided in binary form under this license must not be reverse
+ *    engineered, decompiled, modified and/or disassembled.
+ *
+ * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
+ * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* eslint-disable no-underscore-dangle */
+
+import fs from 'fs';
+import { deserialize } from 'bson';
+import { createInflateRaw } from 'zlib';
+import { Writable } from 'stream';
+
+const setupBuffer = async filename => {
+    const buffer = Buffer.alloc(370 * 1e6);
+    let size = 0;
+    const content = new Writable({
+        write(chunk, _encoding, callback) {
+            chunk.copy(buffer, size, 0);
+            size += chunk.length;
+            callback();
+        },
+    });
+
+    await new Promise(resolve =>
+        fs
+            .createReadStream(filename)
+            .pipe(createInflateRaw())
+            .pipe(content)
+            .on('finish', resolve)
+    );
+    return buffer.slice(0, size);
+};
+
+export default () => ({
+    buffer: null,
+    pos: 0,
+    len: null,
+
+    async initialise(filename) {
+        this.buffer = await setupBuffer(filename);
+        this._moveBufferPointer();
+    },
```

Comment:
Another aspect: At least currently you are calling the `const fileLoader = loadData()`, `await fileLoader.initialise(filename)`, and `fileLoader.load()` all together. So it might as well be just a single function call as long as you do not want to be able to split up the calls. Or was that your intention, e.g. to test them separately? 

An alternative to moving to a class as Bence suggested (and which would enable you to use proper private methods and fields) you could also just export this single function I mention and pass the relevant things around, which would also keep all functions “private” that are just for internal use, since they are not exported. Give me a moment, I will sketch a variant of what I mean.

Reply:
This is what I thought of: https://github.com/NordicSemiconductor/pc-nrfconnect-ppk/compare/update/add-savefile-version...experiment/file_loader. Please take it with a grain of salt, I did not test it just quickly hacked it together, so there may be dragons. 

Besides exposing just a single function from the file, I also did some more changes about things that confused me: 
- Removes the duplicated code of how to slice the buffer and move the buffer pointer. 
- Removes the split between reading a chunk and moving the buffer pointer, which included reading the length of the next chunk.
- Move logic about reading metadata for legacy files from the main function to `loadMetadata`.
- Move logic whether to bits data can be read into `loadBits`.

I did not take a look at `saveFileHandler.js`, but I assume it could also be restyled in a similar fashion. If you would like to, could also take a stab at that.

Feel free to use any parts of this that you like or also just dispose it if you think this makes things worse.

Reply:
I like these changes! :)